lib/galaxy/workflow/refactor/execute.py

```diff
@@ -0,0 +1,363 @@
+        step = self._step_with_module(input_step_dict["id"])
+        module = step.module
+        inputs = module.get_all_inputs()
+
+        input_def = None
+        found_input_names = []
+        for input in inputs:
+            found_input_name = input["name"]
+            found_input_names.append(found_input_name)
+            if found_input_name == input_name:
+                input_def = input
+                break
+        if input_def is None:
+            raise RequestParameterInvalidException(f"Failed to find input with name {input_name} on step {input_step_dict['id']} - input names found {found_input_names}")
+        if input_def.get("multiple", False):
+            raise RequestParameterInvalidException("Cannot extract input for multi-input inputs")
+
+        module_input_type = input_def.get("input_type")
+        # convert dataset, dataset_collection => data, data_collection for refactor API
+        input_type = {
+            "dataset": "data",
+            "dataset_collection": "data_collection",
+        }.get(module_input_type, module_input_type)
+
+        input_action = AddInputAction(
+            action_type="add_input",
+            optional=input_def.get("optional"),
+            type=input_type,
+            label=action.label,
+            position=action.position,
+        )
+        new_input_order_index = self._add_input_get_order_index(input_action)
+        connect_action = ConnectAction(
+            action_type="connect",
+            input=action.input,
+            output=OutputReferenceByOrderIndex(order_index=new_input_order_index),
+        )
+        self._apply_connect(connect_action)
+
+    def _apply_extract_legacy_parameter(self, action: ExtractLegacyParameter):
+        legacy_parameter_name = action.name
+        target_value = "${%s}" % legacy_parameter_name
+
+        target_tool_inputs = []
+
+        for step_def, step in self._iterate_over_step_pairs():
+            module = step.module
+            if module.type != "tool":
+                continue
+
+            tool = module.tool
+            tool_inputs = module.state
+
+            replace_tool_state = False
+
+            def callback(input, prefixed_name, context, value=None, **kwargs):
+                nonlocal replace_tool_state
+                # data parameters cannot have legacy parameter values
+                if input.type in ['data', 'data_collection']:
+                    return NO_REPLACEMENT
+
+                if not contains_workflow_parameter(value):
+                    return NO_REPLACEMENT
+
+                if value == target_value:
+                    target_tool_inputs.append((step.order_index, input, prefixed_name))
+                    replace_tool_state = True
+                    return runtime_to_json(ConnectedValue())
+                else:
+                    return NO_REPLACEMENT
+            visit_input_values(tool.inputs, tool_inputs.inputs, callback, no_replacement_value=NO_REPLACEMENT)
```

Comment:
But that is fine right because with d7e178eca1479c3830526d85adea9c3075f8392e the parameters will work the same way and the tool form will only render the new version.